### PR TITLE
New tests: goldens; unified benchmark+examples

### PR DESCRIPTION
**Pull request by srid**
- [x] Use sydtest, and start porting ./examples
- [ ] Finish rest of ./examples
- [ ] Port benchmarks to ./plutarch-test
- [ ] Delete examples and benchmarks (but keep `plutarch-benchmark` library)
- [x] Make plutarch-test library

### CODE REVIEW

**Comment by srid** [comment]
@L-as What are your thoughts on merging this right now -- so that when people open new PRs with tests, they can start adding the tests to ./plutarch-tests instead -- with the rest of the tasks done as separate PR?

https://github.com/Plutonomicon/plutarch/issues/269#issuecomment-1036411927

**Comment by MangoIV** [comment]
I'm so sorry, I accidentally pushed to your branch

**Comment by srid** [comment]
Ran CI locally, and it passes.

@L-as I've deleted already ported over code. Let's merge this now, and ask current/ future PR authors to add new tests to `./plutarch-test` going forward.

**Review by MatthewCroughan — APPROVED** [review]
Once @NorfairKing provides a flake for his projects, we can reduce the complexity of this code.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:1]: lol

File: plutarch-test/src/Spec.hs
Comment (L-as) [inline plutarch-test/src/Spec.hs:1]: What does this do?

**Review by L-as — DISMISSED** [review]
Looks great, thanks!

**Comment by L-as** [comment]
@srid @MangoIV there are conflicts...

File: plutarch-test/src/Spec.hs
Comment (MangoIV) [reply plutarch-test/src/Spec.hs:1]: Test discovery, afaik

File: plutarch-test/src/Plutarch/Test.hs
Comment (NorfairKing) [inline plutarch-test/src/Plutarch/Test.hs:133]: I would use `sydtest-aeson`'s function for this.

File: plutarch-test/src/Main.hs
Comment (L-as) [reply plutarch-test/src/Main.hs:9]: Inside builds there is no locale IIRC

File: plutarch-test/src/Plutarch/Test.hs
Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:133]: Some goldens contain multiple entries, eg:

```
❯ cat plutarch-test/goldens/bool.pnot.bench.golden 
lam {"exBudgetCPU":29873,"exBudgetMemory":200,"scriptSizeBytes":14}
app {"exBudgetCPU":387150,"exBudgetMemory":1401,"scriptSizeBytes":16}
```

Looks like `sydtest-aeson` shows a JSON-friendly diff. Can we get that for the above case? `pureGoldenJSONValueFileMany :: (Show a, Eq a, FromJSON a, ToJSON a) => FilePath -> Map k a -> GoldenTest a`?

Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:133]: The reason why some goldens have multiple JSON values in them is to group benchmarks for closely related programs in the same file, and thus avoid having gazillion golden files most of which corresponding to tiny programs.

**Comment by srid** [comment]
CI is still down; but tests pass locally.

File: plutarch-test/src/Plutarch/Test.hs
Comment (NorfairKing) [reply plutarch-test/src/Plutarch/Test.hs:131]: @srid if we have the name for each of those things in the multi-golden, we could make a json object out of them, otherwise a list...

Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:131]: Using json object of assoc list (`[String, a]`), `sydtest-aeson` is producing:

image

compared to the current:

image

I don't like how the latter (ie., current behaviour) puts both the entries (`app` and `lam`) in the _same_ line; at the same time I find the former to be too verbose. 🤔

Comment (NorfairKing) [reply plutarch-test/src/Plutarch/Test.hs]: I like the `pureGoldenTextFile` one best. But it's going to be messed up if we change the keys to a different name. Up to you.

Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs]: Keys and order shouldn't change unless developers purposefully change their name/order in the test source. The above golden is for:

```haskell
      goldens
        All
        [ ("lam", popaque pnot)
        , ("app", popaque $ pnot #$ pcon PTrue)
        ]
```

**Comment by L-as** [comment]
So there are "golden" files for the generated UPLC, but there should be two such generated outputs: pre-evaluation term and post-evaluation term.

**Comment by srid** [comment]
> So there are "golden" files for the generated UPLC, but there should be two such generated outputs: pre-evaluation term and post-evaluation term.

#285

**Review by nixinator — COMMENTED** [review]
wow,  this is even bigger ... Srid now wins a cookie, for the biggest commit i've ever seen! This amazing!

**Comment by L-as** [comment]
nitpick: I think uplc-eval is better than uplc.eval, otherwise it's harder to mentally parse the file name IMO. TBF you can also keep it like this.

File: plutarch-test/src/Main.hs
Comment (DarthPJB) [reply plutarch-test/src/Main.hs:9]: I don't actually know, I would however imagine this is a safeguard against GB18030 encoding on Chinese systems; most Chinese distros set it as the default at kernel-level; and it trickles down - I've seen more than my fair share of software break due to systems expecting UTF-8 being force-fed half-byte-codes.

a statement of reality is that all software is locale dependant once it starts manipulating byte-strings, it's just hand-waved by generalised UTF-8 usage.

File: cabal.project
Comment (DarthPJB) [inline cabal.project:3]: I don't understand cabal properly yet; however wildcards always make me nervous.

Are there any implications here for future additions to 'plutarch' causing unexpected behaviour?

**Review by DarthPJB — APPROVED** [review]
I cannot obviously vouch for the density of haskell here - so I've viewed it, read it, become confused, and assumed it's sane and workable.

This review relates only to Nix, besides one question:

why is the file "[plutarch-test/goldens/list.pmatch.bench.golden](https://github.com/Plutonomicon/plutarch/pull/276/files#diff-b70e9645e601fdb41fb7878149692963e431a0935826f1d37fd912a4b5af8932)" empty? I'd expected something, unless it's just a placeholder?

As far as the nix-code goes, it looks to my eye well ordered, and sane - while working around issues with the Haskell ecosystem in an effectively commented fashion. As such, I approve of it.

That said, I'd defiantly hope that @MatthewCroughan will pass his eye over it before merge; as he is without question more likely to find issues - if they exist.

File: cabal.project
Comment (srid) [reply cabal.project:3]: Using wildcards in cabal.project is not uncommon. In fact the default value already uses one. See https://cabal.readthedocs.io/en/3.6/cabal-project.html#cfg-field-packages

**Comment by srid** [comment]
> why is the file "[plutarch-test/goldens/list.pmatch.bench.golden](https://github.com/Plutonomicon/plutarch/pull/276/files#diff-b70e9645e601fdb41fb7878149692963e431a0935826f1d37fd912a4b5af8932)" empty? I'd expected something, unless it's just a placeholder?

@DarthPJB Good catch! That got committed by accident. Removed now.

**Comment by L-as** [comment]
The code should preferably err on unused golden files. We can do this in the future, but someone should make an issue after we merge this.

File: flake.nix
Comment (L-as) [inline flake.nix:566]: Isn't this change entirely unrelated? What does this do?

Comment (L-as) [reply flake.nix:566]: OK, I see, it's for setting `-development` to false. You've put a lot of stuff into a single PR which makes it hard to merge.

File: plutarch-test/plutarch-test.cabal
Comment (L-as) [inline plutarch-test/plutarch-test.cabal:89]: Do you use this?

File: plutarch-test/plutarch-test.cabal
Comment (L-as) [inline plutarch-test/plutarch-test.cabal:114]: Why are you using the same prefix as `Plutarch`?

File: flake.nix
Comment (srid) [reply flake.nix:566]: `haddockProject` is now used only for building haddocks (this rename makes that clear); the rest of the things use `flakeMatrix` (enumerates on 'development' flag as well as ghc version). All the Nix changes are all done to improve test coverage (across flag valus and compiler versions). I don't see any more Nix changes to be done in this branch.

I wanted 8.10 and development=true to be enabled in CI first, before basing other work on. Would you rather I open subsequent PRs (for porting existing tests) as separate PRs, instead of pushing to this branch?

**Comment by L-as** [comment]
What do you do to handle the discrepancy in the strings passed to `fails` in GHC 8.10 and GHC 9.2?  To me it seems like a good solution might just be to use a placeholder string in `fails` when a `testing` flag is enabled.

File: plutarch-test/plutarch-test.cabal
Comment (srid) [reply plutarch-test/plutarch-test.cabal:114]: Mostly existing convention; `Foo.hs` vs `FooSpec.hs`. Example:

- https://github.com/NorfairKing/validity/blob/master/genvalidity-time/src/Data/GenValidity/Time.hs
- https://github.com/NorfairKing/validity/blob/master/genvalidity-time/test/Data/GenValidity/TimeSpec.hs

Comment (L-as) [reply plutarch-test/plutarch-test.cabal:114]: OK, makes sense.

**Comment by srid** [comment]
> What do you do to handle the discrepancy in the strings passed to fails in GHC 8.10 and GHC 9.2?

This PR doesn't provide any general solution to this discrepancy. It sidesteps it entirely (for now), by rewriting that one example to handle the pmatch case exhaustively using perror, https://github.com/Plutonomicon/plutarch/pull/276/commits/e4245abb3603b34d454ee45a45dc16c10d685dbd#diff-3a97af4dc66f7003646b7980f30c5feefff6069a0f5857fa9bb9c1da173f7a3cL182-R193

> To me it seems like a good solution might just be to use a placeholder string in fails when a testing flag is enabled.

In principle it seems to be a bad idea to change application's behaviour[^1] differently when run in tests from when it is used in other contexts. cc @NorfairKing 

[^1]: Specifically, by changing the string here to some constant if run in tests only (but using existing behaviour when used by the user): https://github.com/Plutonomicon/plutarch/blob/a0cbe99921aad7c5df9239cb0240933e4d9b2eaa/Plutarch/TermCont.hs#L39-L41

**Comment by L-as** [comment]
You're right wrt. that. But with this approach we don't test the monadic code properly.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:129]: You should remove the `forall`, it's not necessary here.

Comment (L-as) [reply plutarch-test/src/Plutarch/Test.hs:129]: Same everywhere else where the types are simple.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:2]: I thought you wanted to use GHC 8.10 with this.

File: plutarch-test/src/Plutarch/PIsDataSpec.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/PIsDataSpec.hs:93]: Naming. `pdataCompat` and such.

File: plutarch-test/src/Plutarch/MonadicSpec.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/MonadicSpec.hs:10]: For the future, please try to avoid implicit imports like this. Since this isn't important code, we shouldn't block the PR on this though.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:44]: Please use explicit imports, `(..)` is also implicit.

File: plutarch-test/goldens/unit.pmatch.bench.golden
Comment (L-as) [inline plutarch-test/goldens/unit.pmatch.bench.golden:1]: No `uplc` results for this?

File: plutarch-test/goldens/trace.dev=false.ptrace.uplc.eval.golden
Comment (L-as) [inline plutarch-test/goldens/trace.dev=false.ptrace.uplc.eval.golden:1]: How do you decide which tests to make versions of for both `dev=true` and `dev=false`?

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:116]: Ah, you use this to choose which to test twice?

File: plutarch-test/src/Plutarch/BoolSpec.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/BoolSpec.hs:1]: This isn't your fault, but GH doesn't syntax-highlight this file for some reason...

File: plutarch-test/src/Plutarch/BoolSpec.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/BoolSpec.hs:42]: While not a blocking issue, I do wonder how we could make the testing API more ergonomic.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [inline plutarch-test/src/Plutarch/Test.hs:175]: Perhaps this should be `ClosedTerm POpaque` for the sake of consistency and type inference? Your choice.

**Review by L-as — CHANGES_REQUESTED** [review]
Looks mostly great, thanks!

File: flake.nix
Comment (L-as) [reply flake.nix:523]: A good example is `hsPkgs`.

File: plutarch-test/src/Plutarch/MonadicSpec.hs
Comment (L-as) [reply plutarch-test/src/Plutarch/MonadicSpec.hs:10]: Note that I do know that the tests originally had this problem due to laziness on my part.

File: plutarch-test/src/Plutarch/Test.hs
Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:175]: The `a` requires ImpredicativeTypes. Keeping it polymorphic is nice, because it avoids popaque in most golden groups, eg:

image

File: plutarch-test/src/Plutarch/Test.hs
Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:2]: It is needed for the `a` in `goldens :: PlutarchGolden -> [(String, ClosedTerm a)] `

File: plutarch-test/src/Plutarch/Test.hs
Comment (srid) [reply plutarch-test/src/Plutarch/Test.hs:116]: Yup, the idea is to use `plutarchDevFlagDescribe` somewhere in test hierarchy where you know that the child sub-tree of tests also use Plutarch programs that will behave differently across 'development' flag sets (eg: with tracing). I'll add this to the README.

File: plutarch-test/src/Plutarch/BoolSpec.hs
Comment (srid) [reply plutarch-test/src/Plutarch/BoolSpec.hs:42]: I'll think about this outside this PR. Maybe some sort of MonadWriter that collects programs on the goal, and batches golden tests at the end ...

File: plutarch-test/src/Plutarch/BoolSpec.hs
Comment (srid) [reply plutarch-test/src/Plutarch/BoolSpec.hs:1]: 🤷🏿 
image

File: flake.nix
Comment (srid) [reply flake.nix:523]: I added `projectMatrix`. I suppose this would be useful from `nix repl`.

File: plutarch-test/goldens/trace.dev=false.ptrace.uplc.eval.golden
Comment (srid) [reply plutarch-test/goldens/trace.dev=false.ptrace.uplc.eval.golden:1]: Answered elsewhere, and explained in README.

File: plutarch-test/goldens/unit.pmatch.bench.golden
Comment (srid) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: @L-as Some tests specifically generate _only_ bench goldens (_or_ only uplc goldens), rather than both, as controlled by this type:

https://github.com/Plutonomicon/plutarch/blob/1ebeff5b2dcf5205c753a55b6efe430b2c7f5b00/plutarch-test/src/Plutarch/Test.hs#L135-L149

That's what is happening here, because of `golden Bench p` (instead of `golden All p`). Should we get rid of `Bench` and *always* generate printTerm regardless of the program? Likewise, are there cases where avoiding benchmarks make sense? It is used in a handful of cases, like:

```
plutarch-test/src/Plutarch/PLamSpec.hs
12:      golden PrintTerm $ plam (\x -> x)
14:      golden PrintTerm $ plam (\_ y -> y)
16:      golden PrintTerm $ plam (\x _ -> plet x $ \_ -> perror)
```

I figured we don't need to benchmark lambdas, inasmuch as the body of a lambda won't be evaluated anyway. But may be it is worth benchmarking the script size anyway.

File: plutarch-test/src/Plutarch/MonadicSpec.hs
Comment (srid) [reply plutarch-test/src/Plutarch/MonadicSpec.hs]: Sure. I assume we can relax on `Plutus.V1.Ledger.Api` and the like for testing code. Added explicit imports for Plutarch.*.

File: plutarch-test/goldens/api.example.getFields.uplc.golden
Comment (srid) [reply plutarch-test/goldens/api.example.getFields.uplc.golden:1]: Added in 48f439aea4e26213c834bf2eba0ce969f732e35f - but see comment below for larger picture.

File: plutarch-test/goldens/unit.pmatch.bench.golden
Comment (srid) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: Here's an example of something generating bench goldens, but no printTerm golden:

```haskell
    describe "phexByteStr" $ do
      let a :: [String] = ["42", "ab", "df", "c9"]
          p = pconstant @PByteString (BS.pack $ map readByte a) #== phexByteStr (concat a)
      golden Bench p
      it "relation" $ passert p
```

At least the eval uplc would be useless here, as it would be: `(program 1.0.0 True)`. The non-eval UPLC would be: `(program 1.0.0 ((\i0 -> equalsByteString i1 i1) #42abdfc9))` - however why should it be under "phexByteStr" tree (see above), rather a tree for equalsByteString?

Comment (srid) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: Here, we can do benchmark golden, because doing that will result in error:

```haskell
    describe "pmatch" $ do
      let p = pmatch (integerList [1, 3, 1]) $ \_ -> perror
      golden PrintTerm p
```

Changing `PrintTerm` -> `All` will result in:

image

Comment (srid) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: Another example, where we can't really benchmark a scriptcontext constant:

```haskell
golden PrintTerm ctx
```
image

Comment (srid) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: Anyway, uplc goldens added in [a001bde](https://github.com/Plutonomicon/plutarch/pull/276/commits/a001bdedb021b4314f3bebad570538ea39cfe4c0)

**Comment by srid** [comment]
@L-as This is ready for merge now. Opened #289 and #290 as follow-up (in addition to porting the rest).

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [reply plutarch-test/src/Plutarch/Test.hs:2]: That is true and is quite unfortunate. You can have a newtype wrapper for `ClosedTerm` if you want, i.e.
```haskell
newtype MyClosedTerm a = MyClosedTerm (ClosedTerm a)
```
which should fix it AFAICT, and then you could do `MyClosedTerm $ ...` in the list you pass to `goldens`. You could also hard-code it to be `POpaque`. It's quite unfortunate that we still support GHC 8.10, but we will probably drop it once GHC 9.2.2 comes out, since I've realised that there isn't much point to the older GHC 8.10 support.

File: plutarch-test/goldens/unit.pmatch.bench.golden
Comment (L-as) [reply plutarch-test/goldens/unit.pmatch.bench.golden:1]: > At least the post-eval uplc would be useless here, as it would be: (program 1.0.0 True). The pre-eval UPLC would be: (program 1.0.0 ((\i0 -> equalsByteString i1 i1) #42abdfc9)) - however why should it be under "phexByteStr" tree (see above), rather a tree for equalsByteString?

I don't agree with this. the post-eval UPLC is still very important, because if it starts evaluating into something else, we have a bug! 9/10 times changes to the golden files are accidental and means we have a bug.

I don't see why we wouldn't support only `All`. Terms that err can still be benchmarked. It just means that the benchmarking result should include only the script size, and possibly still CPU and memory cost if that information isn't lost.

Wrt. the `ctx` case, I'm not sure why that's erring in the first place, but I think that we should still benchmark this too.

You could potentially include the error message *in* the benchmarking golden file.

File: plutarch-test/src/Plutarch/Test.hs
Comment (L-as) [reply plutarch-test/src/Plutarch/Test.hs:171]: Good point.

**Comment by L-as** [comment]
@srid 

I really really think we should use CPP to check for a testing flag in calls to `fails` and just substitute it with a placeholder string in those cases. As you've noted in #290 , even just moving lines around will cause the golden files to change. If a PR isn't buggy, then most often **no golden files should change**. This violates that principle and means that we'll have many unrelated changes to the golden files that don't indicate a bug.

**Review by L-as — CHANGES_REQUESTED** [review]
See previous comment wrt. #290 

Removing everything but `All` would also be nice if you have time.

**Comment by L-as** [comment]
Thanks for the work BTW, this is really great and pretty close to as good as it could get AFAICT. For the future, I think we should try to see if we can group the tests more so we don't have ~200 golden files, albeit that should be another PR.

**Pull request merged**
